// nativescript/lib/commands/platform-clean.js
"use strict";
var CleanCommand = (function () {
    function CleanCommand($platformService, $errors) {
        this.$platformService = $platformService;
        this.$errors = $errors;
        this.allowedParameters = [];
    }
    CleanCommand.prototype.execute = function (args) {
        var _this = this;
        return (function () {
            _this.$platformService.removePlatforms(args).wait();
            _this.$platformService.addPlatforms(args).wait();
        }).future()();
    };
    CleanCommand.prototype.canExecute = function (args) {
        var _this = this;
        return (function () {
            if (!args || args.length === 0) {
                _this.$errors.fail("No platform specified. Please specify a platform to clean");
            }
            _.each(args, function (arg) { return _this.$platformService.validatePlatform(arg); });
            return true;
        }).future()();
    };
    return CleanCommand;
}());
exports.CleanCommand = CleanCommand;
$injector.registerCommand("platform|clean", CleanCommand);
