// nativescript/lib/common/appbuilder/providers/appbuilder-livesync-provider-base.js
"use strict";
var Future = require("fibers/future");
var AppBuilderLiveSyncProviderBase = (function () {
    function AppBuilderLiveSyncProviderBase($androidLiveSyncServiceLocator, $iosLiveSyncServiceLocator) {
        this.$androidLiveSyncServiceLocator = $androidLiveSyncServiceLocator;
        this.$iosLiveSyncServiceLocator = $iosLiveSyncServiceLocator;
    }
    Object.defineProperty(AppBuilderLiveSyncProviderBase.prototype, "deviceSpecificLiveSyncServices", {
        get: function () {
            var _this = this;
            return {
                android: function (_device, $injector) {
                    return $injector.resolve(_this.$androidLiveSyncServiceLocator.factory, { _device: _device });
                },
                ios: function (_device, $injector) {
                    return $injector.resolve(_this.$iosLiveSyncServiceLocator.factory, { _device: _device });
                }
            };
        },
        enumerable: true,
        configurable: true
    });
    AppBuilderLiveSyncProviderBase.prototype.preparePlatformForSync = function (platform) {
        return Future.fromResult();
    };
    AppBuilderLiveSyncProviderBase.prototype.canExecuteFastSync = function (filePath) {
        return false;
    };
    AppBuilderLiveSyncProviderBase.prototype.transferFiles = function (deviceAppData, localToDevicePaths, projectFilesPath, isFullSync) {
        return deviceAppData.device.fileSystem.transferFiles(deviceAppData, localToDevicePaths);
    };
    return AppBuilderLiveSyncProviderBase;
}());
exports.AppBuilderLiveSyncProviderBase = AppBuilderLiveSyncProviderBase;
