// nativescript/lib/common/appbuilder/mobile-platforms-capabilities.js
"use strict";
var MobilePlatformsCapabilities = (function () {
    function MobilePlatformsCapabilities($errors) {
        this.$errors = $errors;
    }
    MobilePlatformsCapabilities.prototype.getPlatformNames = function () {
        return _.keys(this.getAllCapabilities());
    };
    MobilePlatformsCapabilities.prototype.getAllCapabilities = function () {
        this.platformCapabilities = this.platformCapabilities || {
            iOS: {
                wirelessDeploy: true,
                cableDeploy: true,
                companion: true,
                hostPlatformsForDeploy: ["win32", "darwin"]
            },
            Android: {
                wirelessDeploy: true,
                cableDeploy: true,
                companion: true,
                hostPlatformsForDeploy: ["win32", "darwin", "linux"]
            },
            WP8: {
                wirelessDeploy: true,
                cableDeploy: false,
                companion: true,
                hostPlatformsForDeploy: ["win32"]
            }
        };
        return this.platformCapabilities;
    };
    return MobilePlatformsCapabilities;
}());
exports.MobilePlatformsCapabilities = MobilePlatformsCapabilities;
$injector.register("mobilePlatformsCapabilities", MobilePlatformsCapabilities);
