// plist-merge-patch/src/index.js
var plist = require("plist");
var jsonmergepatch = require("json-merge-patch");
var PlistSession = (function () {
    function PlistSession(console) {
        this.console = console;
        this.patches = [];
    }
    PlistSession.prototype.patch = function (patch) {
        this.patches.push(patch);
    };
    PlistSession.prototype.build = function () {
        var _this = this;
        this.log("Start");
        var jsonPlist;
        jsonPlist = {};
        if (this.patches) {
            this.patches.forEach(function (patch) {
                _this.log("Patch '" + patch.name + "'");
                var patchString = patch.read();
                var patchJson = plist.parse(patchString);
                jsonPlist = jsonmergepatch.apply(jsonPlist, patchJson);
            });
        }
        var resultString = plist.build(jsonPlist);
        this.log("Complete");
        return resultString;
    };
    PlistSession.prototype.log = function (msg) {
        if (this.console && this.console.log) {
            this.console.log(msg);
        }
    };
    return PlistSession;
})();
exports.PlistSession = PlistSession;
